### DISCUSSION TITLE
Remove cancellablesLock?

### DISCUSSION THREAD

**Discussion by rcarver** [pointfreeco/swift-composable-architecture#857] [Ideas]
I was tracking down some performance issues that turned out to be caused by an excessive number of calls to `cancel`. That was on my end. But it led me to looking at the source around `.cancel` and `.cancellable` and its use of `NSRecursiveLock`. 

https://github.com/pointfreeco/swift-composable-architecture/blob/14379519edb3ddeb95c3262ad3d808d42c146af9/Sources/ComposableArchitecture/Effects/Cancellation.swift#L41

With the new clarity around thread checking, is this lock still needed? With my app, I did a test of removing the lock and confirmed that everything worked and that that all accesses of `cancellationCancellables` were on the main thread (already true in all but one case, and was easy to fix.)

So, are there reasons to keep the lock? If not, can it be removed and thread checking added to effect cancellation?

**Comment by iampatbrown** [comment]
Hey @rcarver, did you see much of a performance difference when you removed the lock? I have a feeling `PassthroughSubject` is also going to be doing some sort of locking so you'd probably have to replace that too. 

You should be able to make a standalone extension responsible for cancellations as well for when you need more control. In the past I've used something like the following:
```swift
public final class EffectCanceller {
  var cancellables: [AnyHashable: Set<AnyCancellable>] = [:]

  public init() {}

  public func register<P: Publisher>(
    _ publisher: P,
    id: AnyHashable,
    cancelInFlight: Bool = false
  ) -> Effect<P.Output, P.Failure> {
  ...  
  }
}

extension Effect {
  public func cancellable(
    with canceller: EffectCanceller,
    id: AnyHashable,
    cancelInFlight: Bool = false
  ) -> Effect {
    canceller.register(self, id: id, cancelInFlight: cancelInFlight)
  }
}
```
Here's a full example if you want to use it as a starting point  [iampatbrown:effect-canceller](https://github.com/iampatbrown/swift-composable-architecture/blob/a191f6a401684c1fea3101eb287d27b9c88a18b5/Sources/ComposableArchitecture/Effects/Cancellation.swift#L100)

**Reply by andreyz** [reply]
Could you give an example use case of when did above cancellation extension was useful?

**Reply by iampatbrown** [reply]
I was mainly experimenting with the ideas discussed in #335 and #345. Ultimately though, I tend to use more explicit cancellations and something like the [higher order lifecycle reducer](https://github.com/pointfreeco/swift-composable-architecture/blob/main/Examples/CaseStudies/SwiftUICaseStudies/04-HigherOrderReducers-Lifecycle.swift) when working with things like `NavigationView` and `IfLetStore`.

The custom `EffectCanceller` would outperform the global cancellations in benchmarks but I didn't think it was enough to justify the added complexity. 

@rcarver when I tested it a while ago I was either using [swift-benchmark](https://github.com/google/swift-benchmark) or putting small examples in loops and using the Time Profiler.

**Comment by rcarver** [comment]
Thanks @iampatbrown. Unfortunately haven't been able to confirm any performance difference without the lock. I've tried a few things but it may be beyond my Instruments skills. Any ideas of how to do it?

Your extension is handy - it can make cancellation more testable/observable at least. 

@andreyz I can't speak to the original motivation but there's two possibilities: optimizing for a specific use case (performance) and testability.

**Reply by rcarver** [reply]
@iampatbrown I haven't benchmarked it yet, but here's what I came up with using your code as a start.

https://github.com/rcarver/swift-composable-presentation/pull/1

That's building on this
https://github.com/darrarski/swift-composable-presentation/pull/4

Overall I like this approach to presentation, the problem is that you accumulate`.cancellable` for every effect and level of presentation. My app has recursive infinite navigation so I'm pretty sure this is adding up.

**Reply by iampatbrown** [reply]
@rcarver looks pretty cool. I ended up doing something similar to `@Presented` as well. Didn't get it working quite how I wanted though. Will probably revisit it at some point. 

I'll give your implementation a spin in the next couple of days. I'm pretty sure TCA uses `Deferred` when registering effects for cancellation to prevent any work from being performed prematurely. Not sure if this is needed in your case though, just something to keep in mind. 

Would it be fairly straight forward for me to update the `swift-composable-presentation` example using your wrapper?

**Reply by rcarver** [reply]
Thanks! Good call on Deferred, that was probably cutting too far.

> Would it be fairly straight forward for me to update the swift-composable-presentation example using your wrapper?

I think so, should just be adding the property wrapper. I should add that to the PR actually. 

I have another more complex navigation example I'm working on too, will link that in the next few days.

**Reply by iampatbrown** [reply]
> Thanks! Good call on Deferred, that was probably cutting too far.

I did the exact same thing :P 

> I have another more complex navigation example I'm working on too, will link that in the next few days.

Oh awesome. I might wait for that then :)

**Reply by stephencelis** [reply]
If a measurable performance difference can be detected, we'd definitely be open to changing this! We were under the impression, though, that locking/cancellation code isn't invoked in hot enough code paths for it to introduce performance problems. Please let us know, though, if you've found to the contrary!

**Reply by rcarver** [reply]
Thanks @stephencelis - hopefully I'll have some time to benchmark it soon.

It's a hot path when using approach to automatic presentation and cancellation in [Composable Presentation](https://github.com/darrarski/swift-composable-presentation) - the [key function is here](https://github.com/darrarski/swift-composable-presentation/blob/main/Sources/ComposablePresentation/Reducer%2BCombined.swift)

In this approach, each action and effect in the "presented" state is wrapped by `.cancellable`. Even worse, any nested presented states are *also* wrapped. I'm working on some specific optimizations to that library to reduce unnecessary calls to cancellable/cancel but the core issue still remains.

FWIW, this has been the best approach to navigation in my app but I'm all ears at better ways to model navigation overall.

**Reply by stephencelis** [reply]
> It's a hot path when using approach to automatic presentation and cancellation in [Composable Presentation](https://github.com/darrarski/swift-composable-presentation) - the [key function is here](https://github.com/darrarski/swift-composable-presentation/blob/main/Sources/ComposablePresentation/Reducer%2BCombined.swift)

I think even in this case, the lock will only run whenever child effects complete. So while every child effect will invoke the lock, I don't think much time will be spent there.

Even if a child feature ran a timer on a CA display link and emitted an effect that completed every frame, I'm not sure that would pose a problem, though that's maybe the first stress test I'd try out!

**Comment by rcarver** [comment]
As promised! Here's the extended example of recursive navigation and presentation:
https://github.com/rcarver/tca-example-recursive-navigation

I also updated the Composable Presentation Example app [in my fork](https://github.com/rcarver/swift-composable-presentation/pull/1) in [this commit](https://github.com/rcarver/swift-composable-presentation/pull/1/commits/ab312fec336caf2a1ef5abc55ca22ad3508daf6a).
